OpenRA.Mods.Common/Traits/Buildings/Gate.cs

```diff
@@ -0,0 +1,133 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Will open and be passable for actors that appear friendly when there are no enemies in range.")]
+	public class GateInfo : UpgradableTraitInfo, ITraitInfo, Requires<RenderSimpleInfo>
+	{
+		public readonly string OpenSequence = "open";
+		public readonly string ClosingSequence = "closing";
+		public readonly string ClosedSequence = "closed";
+
+		public readonly string OpeningSound = null;
+		public readonly string ClosingSound = null;
+
+		[Desc("`-` means it blocks when closed.")]
+		public readonly string Footprint = "-";
+		public readonly CVec Dimensions = new CVec(1, 1);
+
+		[Desc("How far to search for allied and enemy units.")]
+		public readonly WRange ScanRange = WRange.FromCells(3);
+
+		public object Create(ActorInitializer init) { return new Gate(init, this); }
+	}
+
+	public class Gate : ITick, IOccupySpace, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly GateInfo info;
+		readonly Actor self;
+		readonly RenderSimple renderSimple;
+		readonly CPos topLeft;
+
+		WPos center;
+		bool cachedOpen;
+		Pair<CPos, SubCell>[] blockedCells;
+
+		public Gate(ActorInitializer init, GateInfo info)
+		{
+			this.info = info;
+			this.self = init.Self;
+
+			renderSimple = self.Trait<RenderSimple>();
+
+			topLeft = init.Get<LocationInit, CPos>();
+			center = init.World.Map.CenterOfCell(topLeft) +
+				((init.World.Map.CenterOfCell(CPos.Zero + new CVec(info.Dimensions.X, info.Dimensions.Y))
+					- init.World.Map.CenterOfCell(new CPos(1, 1))) / 2);
```

Comment:
This is duplicated uglyness from FootprintUtils.  What exactly was the problem with making Gate subclass Building?  It would be much cleaner if you did, and then used FootprintUtils instead.